src/main/java/com/conveyal/datatools/manager/jobs/MergeFeedsJob.java

```diff
@@ -375,17 +405,25 @@ private int constructMergedTable(Table table, List<FeedToMerge> feedsToMerge,
         try {
             // Get shared fields between all feeds being merged. This is used to filter the spec fields so that only
             // fields found in the collection of feeds are included in the merged table.
-            Set<Field> sharedFields = getSharedFields(feedsToMerge, table);
-            // Initialize future feed's first date to the first calendar date from the validation result.
+            Set<Field> allFields = getAllFields(feedsToMerge, table);
+            // Initialize future and active feed's first date to the first calendar date from validation result.
             // This is equivalent to either the earliest date of service defined for a calendar_date record or the
             // earliest start_date value for a calendars.txt record. For MTC, however, they require that GTFS
             // providers use calendars.txt entries and prefer that this value (which is used to determine cutoff
             // dates for the active feed when merging with the future) be strictly assigned the earliest
             // calendar#start_date (unless that table for some reason does not exist).
-            LocalDate futureFeedFirstDate = feedsToMerge.get(0).version.validationResult.firstCalendarDate;
+            LocalDate futureFeedFirstDate = getFutureFeed().version.validationResult.firstCalendarDate;
+            LocalDate activeFeedFirstDate = getActiveFeed().version.validationResult.firstCalendarDate;
             LocalDate futureFirstCalendarStartDate = LocalDate.MAX;
             // Iterate over each zip file. For service period merge, the first feed is the future GTFS.
             for (int feedIndex = 0; feedIndex < feedsToMerge.size(); feedIndex++) {
+                boolean handlingActiveFeed = feedIndex > 0;
+                boolean handlingFutureFeed = feedIndex == 0;
+                if (EXTEND_FUTURE.equals(mergeFeedsResult.mergeStrategy) && handlingActiveFeed) {
+                    // No need to iterate over second (active) file if strategy is to simply extend the future GTFS
+                    // service to start earlier.
+                    continue;
+                }
```

Comment:
I used the [SolarLint static code analysis tool](https://www.sonarlint.org/) and this `constructMergedTable` method has a cognitive complexity of **390** which is severely beyond the recommended maximum of **15**. Any further additions of logical branches to this method like this one here are making it that much more difficult to read and understand. This file itself is above 1k lines of code. This all needs a major refactor into more manageable and readable methods.